### DISCUSSION TITLE
Add option to override backoffs

### DISCUSSION THREAD

**Discussion by wlach** [hynek/stamina#59] [Ideas]
Stamina has a nice option to disable retrying altogether (as documented in https://stamina.hynek.me/en/latest/tutorial.html#deactivating-retries-globally) but what if you want to actually test retries?

With tenacity we can do this (as seen here: https://stackoverflow.com/a/48425554):

```py
from tenacity import wait_none

func.retry.wait = wait_none()
```

But I don't see an equivalent feature for stamina. Could we add one?

**Comment by hynek** [comment]
I wouldn't be comfortable to add a general "fiddle with setting" solution, because people would immediately start abusing it, but I supposed we could add for this use-case only (I always have a soft spot in my heart for testing) a `stamina.set_no_waiting(True/False)`.

**Reply by tino** [reply]
Yes, I'd love that! I write tests for proper retries, but those now add seconds to my otherwise sub-second test suite.

I'm using:
```python
@pytest.fixture(autouse=True, scope="session")
def no_wait_for_stamina_retries():
    # Set all wait times between retries to 0.
    # Note that this won't work yet on @stamina.retry decorators, as we are too
    # late with patching for it to work.
    with mock.patch("stamina._core._t.wait_exponential_jitter", return_value=0):
        yield
```

but as commented, that doesn't work for function decorators... (And patching privates of a lib is also ugly 😅)

**Comment by hynek** [comment]
@wlach @tino looking at this, are y'all sure that you just want to disable backoff? Because it kinda means that you always get the maximum number of attempts (or even whatever fits into _wait_max_ if none is set)?

I emphatically don't want global configuration, it seems like we want something like `def set_testing(attempts=1)` that disables backoffs and allows to limit the number of attempts + timeout?

**Reply by hynek** [reply]
Work is underway in https://github.com/hynek/stamina/pull/73

**Reply by wlach** [reply]
@hynek Yeah I think the way you did it makes a ton of sense. I've been thinking a bit more about this and I think my use case is mostly that I want to have assertions in my tests that there *is* some kind retrying. As things stand, it would be relatively easy for someone to just remove the stamina decorators (e.g. during an overzealous refactoring) and have tests still pass.

**Reply by hynek** [reply]
That should be doable via instrumentation, I think?

**Comment by hynek** [comment]
This is now fixed by https://github.com/hynek/stamina/pull/73.
